### ISSUE TITLE
mobx-react 6 shallow testing regression

### ISSUE THREAD

**Issue by sergei-startsev** [mobxjs/mobx-react#699]
**What is the current behavior?**

 With `mobx-react` 6 upgrade regressions are detected in unit tests:

```jsx
// Foo.jsx
class Foo extends React.Component {
  //...
  render() {
    return <div>Test</div>;
  }
}
export default inject('store')(observer(Foo));
```
```jsx
//Foo.test.jsx
describe('Foo', () => {
  test('shallow', () => {
    const component = shallow(<Foo.wrappedComponent />);
    expect(component).toMatchSnapshot();
  });
})
```
Snapshot:
```jsx
<Observer>
  <Component />
</Observer>
```
Here's the repository with repro steps: https://github.com/sergei-startsev/mobx-react-6-regression

**What is the expected behavior?**

Expected snapshot:
```jsx
<div>
  Test
</div>
```
---
I believe the issue is caused by added `<Observer>` wrapper:
https://github.com/mobxjs/mobx-react/blob/4904b624f46be2e14d3a3d1017c1fa9bcb334f02/src/observer.js#L108

See https://github.com/mobxjs/mobx-react/commit/1ea40e21117d9c30067dd7b0267aa80b99473c59#diff-08f3e9ab539253c87ea111cc61c69e29R156.

**Comment by FredyC** [comment]
This is one of the documented changes. You have to update your snapshots, it's not a regression.

https://github.com/mobxjs/mobx-react/blob/master/CHANGELOG.md#600

> For observer based components, there will now be an additional Observer component in the tree.

**Comment by sergei-startsev** [comment]
@FredyC These snapshots are useless:
```jsx
<Observer>
  <Component />
</Observer>
```

**Comment by sergei-startsev** [comment]
At the moment shallow rendering doesn't provide any useful results when testing observable components,  only the `<Observer>` is rendered.  So how can I setup shallow testing with recent changes?

**Comment by FredyC** [comment]
Ok sorry, I was too quick on close here. I will leave it open for someone else to chime in.

Personally, I don't use shallow rendering. It's kinda frowned-upon practice anyway. Do yourself (and your team) a service and buy it ... https://testingjavascript.com/

**Comment by sergei-startsev** [comment]
[Enzyme team disagree:](https://airbnb.io/enzyme/docs/api/shallow.html#shallow-rendering-api)
>Shallow rendering is useful to constrain yourself to testing a component as a unit, and to ensure that your tests aren't indirectly asserting on behavior of child components.

Testing a whole component tree makes tests fragile since any changes in child components affect your test while unit testing is about testing a module itself, not its dependencies.

**Comment by FredyC** [comment]
Well, that's why I don't Enzyme 😆 That would be a long talk and it's definitely the wrong place to debate over testing approaches.

For now, I would advise you to stick to V5 unless V6 is solving something else for you.

**Comment by urugator** [comment]
Not familiar with Enzyme, checked the API briefly, would this work?
```javascript
const component = shallow(<Foo.wrappedComponent />).find(Observer).dive();
expect(component).toMatchSnapshot();
```

**Comment by sergei-startsev** [comment]
@urugator 
Yes, you'll get a snapshot of your component, not `Observer` wrapper. Since you can pass [component’s `displayName` as a selector](https://airbnb.io/enzyme/docs/api/selector.html#3-a-react-components-displayname), the following example will also work:

```jsx
const component = shallow(<Foo.wrappedComponent />).find('Observer').dive();
expect(component).toMatchSnapshot();
```
Snapshot:
```jsx
<div>
   Test
</div>
```

But I'm not sure if it's a good idea to let unit tests know anything about the `Observer` and its internal tree.  I'm also not sure that changing existing tests is an option here though...

**Comment by mweststrate** [comment]
I think this issue is another good argument to revert to the old "mixin" based `observer` implementation for classes?

**Comment by mweststrate** [comment]
Fixed by 6.1.0
